docs/advanced.md

```diff
@@ -471,3 +471,160 @@ If you do need to make HTTPS connections to a local server, for example to test
 >>> r
 Response <200 OK>
 ```
+
+## Middleware
+
+Middleware is a general-purpose mechanism for extending the built-in functionality of a `Client`.
+
+### Using middleware
+
+Middleware generally comes in the form of classes. Middleware classes and their configuration parameters are meant to be passed as a list of `httpx.Middleware` instances to a client:
+
+```python
+from example.middleware import ExampleMiddleware
+import httpx
+
+middleware = [
+    httpx.Middleware(ExampleMiddleware, client_param="value", ...),
+]
+
+with httpx.Client(middleware=middleware) as client:
+    # This request will pass through ExampleMiddleware before
+    # reaching the core processing layers of the Client.
+    r = client.get("https://example.org")
+```
+
+### Writing middleware
+
+Middleware classes should accept a `get_response` parameter, that represents the inner middleware, as well as any keyword arguments for middleware-specific configuration options. They should implement the `.__call__()` method, that accepts a `request`, `timeout` configuration, and any per-request keyword arguments coming directly from keyword arguments passed to `client.get()`, `client.post()`, etc. The `__call__()` method should return a generator that returns a `Response` instance. To get a response from the inner middleware, use `response = yield from get_response(...)`.
+
+#### Basic example
+
+Here is a "do-nothing" middleware that sends the request unmodified, and returns the response unmodified:
+
+```python
+import httpx
+
+class PassThroughMiddleware:
+    def __init__(self, get_response):
+        self.get_response = get_response
+
+    def __call__(self, request, timeout, **kwargs):
```

Comment:
It just occurred to me now that we might want to make it so that `timeout` is buried inside `**kwargs`. It’s not useful for the vast majority of middleware, so exposing it as a positional argument in the API is somewhat clunky.